src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.CheckOrReachability.cs

```diff
@@ -0,0 +1,956 @@
+          LabelSymbol defaultLabel,
+          SyntaxNode syntax,
+          BindingDiagnosticBag diagnostics,
+          bool isSwitchExpression = false)
+        {
+            CheckOrReachability(previousCases, patternIndex, pattern,
+                builder, rootIdentifier, defaultLabel, syntax, diagnostics, ignoreDefaultLabel: isSwitchExpression);
+
+            CheckOrReachability(previousCases, patternIndex, MoveNotPatternsDownRewriter.MakeNegatedPattern(pattern),
+                builder, rootIdentifier, defaultLabel, syntax, diagnostics, ignoreDefaultLabel: isSwitchExpression);
+        }
+
+        private static void CheckOrReachability(
+            ArrayBuilder<StateForCase> previousCases,
+            int patternIndex,
+            BoundPattern pattern,
+            DecisionDagBuilder builder,
+            BoundDagTemp rootIdentifier,
+            LabelSymbol defaultLabel,
+            SyntaxNode syntax,
+            BindingDiagnosticBag diagnostics,
+            bool ignoreDefaultLabel = false)
+        {
+            var normalizedPattern = MoveNotPatternsDownRewriter.Rewrite(pattern);
+
+            SetsOfOrCases setOfOrCases = RewriteToSetsOfOrCases(normalizedPattern);
+            if (setOfOrCases.IsDefault)
+            {
+                return;
+            }
+
+            // We construct a DAG and analyze reachability of branches once per `or` sequence
+            Debug.Assert(!setOfOrCases.IsDefault);
+            foreach (OrCases orCases in setOfOrCases.Set)
+            {
+                using var casesBuilder = TemporaryArray<StateForCase>.GetInstance(orCases.Cases.Count);
+                var labelsToIgnore = PooledHashSet<LabelSymbol>.GetInstance();
+                PopulateStateForCases(builder, rootIdentifier, previousCases, patternIndex, orCases, labelsToIgnore, syntax, ref casesBuilder.AsRef());
+                BoundDecisionDag dag = builder.MakeBoundDecisionDag(syntax, ref casesBuilder.AsRef());
+
+                foreach (StateForCase @case in casesBuilder)
+                {
+                    if (!dag.ReachableLabels.Contains(@case.CaseLabel) && !labelsToIgnore.Contains(@case.CaseLabel))
+                    {
+                        diagnostics.Add(ErrorCode.WRN_RedundantPattern, @case.Syntax);
+                    }
+                }
+
+                if (!ignoreDefaultLabel && !dag.ReachableLabels.Contains(defaultLabel))
+                {
+                    diagnostics.Add(ErrorCode.WRN_RedundantPattern, syntax);
+                }
+            }
+        }
+
+        static void PopulateStateForCases(DecisionDagBuilder builder, BoundDagTemp rootIdentifier, ArrayBuilder<StateForCase> previousCases, int patternIndex, OrCases set, PooledHashSet<LabelSymbol> labelsToIgnore,
+            SyntaxNode nodeSyntax, ref TemporaryArray<StateForCase> casesBuilder)
+        {
+            for (int i = 0; i < patternIndex; i++)
+            {
+                casesBuilder.Add(previousCases[i]);
```

Comment:
>previousCases [](http://example.com/codeflow?start=33&length=13)

Does this array actually contain the original cases and ```previousCases``` is actually index of the case we are analyzing right now?

Reply:
Yes, this array contains all the original cases (we compute them just once). We copy all the ones before the pattern/case that we're analyzing right now (which is at index `patternIndex`). Then we add the cases cooked up for current pattern.